hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsInputStream.java

```diff
@@ -224,6 +240,123 @@ private int readOneBlock(final byte[] b, final int off, final int len) throws IO
     return bytesToRead;
   }
 
+  private boolean shouldReadFully() {
+    return this.firstRead && this.context.readSmallFilesCompletely()
+        && this.contentLength <= this.bufferSize;
+  }
+
+  private boolean shouldReadLastBlock(int len) {
+    return this.firstRead && this.context.optimizeFooterRead()
+        && len == FOOTER_SIZE
+        && this.fCursor == this.contentLength - FOOTER_SIZE;
+  }
+
+  private int readFileCompletely(final byte[] b, final int off, final int len)
+      throws IOException {
+    if (closed) {
+      throw new IOException(FSExceptionMessages.STREAM_IS_CLOSED);
+    }
+
+    Preconditions.checkNotNull(b);
+    LOG.debug("read one block requested b.length = {} off {} len {}", b.length,
+        off, len);
+
+    if (len == 0) {
+      return 0;
+    }
+
+    if (this.available() == 0) {
+      return -1;
+    }
+
+    if (off < 0 || len < 0 || len > b.length - off) {
+      throw new IndexOutOfBoundsException();
+    }
+
+    buffer = new byte[bufferSize];
+    // data need to be copied to user buffer from index bCursor, bCursor has
+    // to be the current fCusor
+    bCursor = (int) fCursor;
+    fCursorAfterLastRead = fCursor;
+    int totalBytesRead = 0;
+    int loopCount = 0;
+    // Read from begining
+    fCursor = 0;
+    while (fCursor < contentLength) {
+      int bytesRead = readInternal(fCursor, buffer, limit,
+          (int) contentLength - limit, true);
+      if (bytesRead > 0) {
+        totalBytesRead += bytesRead;
+        limit += bytesRead;
+        fCursor += bytesRead;
+      }
+      if (loopCount++ >= 10) {
```

Comment:
As discussed, this logic needs to be relooked for 
1. What if the requested data is already available in the partial read done ?
2. Reduce the loopCount as the retry logic on ABFS driver can make the client read overall expensive. Fail faster with just 1 or 2 tries.
3. Never fail the read request because optimization code failed to read full file. Fail fast and send read for client requested position.
4. All pointer fields need to be in valid state (bCursor, fCursor, fCursorAfterLastRead). In failure case currently fCursor could be in a different position that last seek done.

Reply:
Made changes the following way
1. The available data is returned
2. Done
3. Done
4. Done